File: javascript/binary-search.js
// Algorithm: Binary Search (Iterative)
// Description: Finds target in sorted array using O(log n) comparisons.
// Usage: Run with `node binary-search.js`.
function binarySearch(data, target) {
  let left = 0;
  let right = data.length - 1;

  while (left <= right) {
    const mid = Math.floor((left + right) / 2);
    if (data[mid] === target) {
      return mid;
    }
    if (data[mid] < target) {
      left = mid + 1;
    } else {
      right = mid - 1;
    }
  }
  return -1;
}

const values = [1, 4, 6, 8, 10, 13];
const target = 8;
const result = binarySearch(values, target);
console.log(result >= 0 ? `Found ${target} at index ${result}` : `${target} not found`);

File: rust/binary_search.rs
// Algorithm: Binary Search (Iterative)
// Description: Locates a target in a sorted slice in O(log n) time using iterative bounds.
// Usage: Run with `rustc binary_search.rs && ./binary_search`.
fn binary_search(data: &[i32], target: i32) -> Option<usize> {
    let mut left: i32 = 0;
    let mut right: i32 = (data.len() as i32) - 1;

    while left <= right {
        let mid = left + ((right - left) / 2);
        let mid_index = mid as usize;
        if data[mid_index] == target {
            return Some(mid_index);
        }
        if data[mid_index] < target {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }

    None
}

fn main() {
    let values = [1, 4, 6, 8, 10];
    let target = 6;
    match binary_search(&values, target) {
        Some(index) => println!("Found {} at index {}", target, index),
        None => println!("{} not found", target),
    }
}
